pkg/internal/controller/controller_test.go

```diff
@@ -122,6 +122,60 @@ var _ = Describe("controller", func() {
 			close(done)
 		})
 
+		It("should error when cache sync timeout occurs", func(done Done) {
+			ctrl.CacheSyncTimeout = 10 * time.Nanosecond
+			c, err := cache.New(cfg, cache.Options{})
+			Expect(err).NotTo(HaveOccurred())
+			_, err = c.GetInformer(context.TODO(), &appsv1.Deployment{})
+			Expect(err).NotTo(HaveOccurred())
+			sync := false
+			ctrl.startWatches = []watchDescription{{
+				src: source.NewKindWithCache(&appsv1.Deployment{}, &informertest.FakeInformers{
+					Synced: &sync,
+				}),
+			}}
+
+			err = ctrl.Start(context.TODO())
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("cache did not sync"))
+
+			close(done)
+		})
+
+		It("should not error when cache sync time out is of reasonable value", func(done Done) {
+			ctrl.CacheSyncTimeout = 1 * time.Second
+			c, err := cache.New(cfg, cache.Options{})
+			Expect(err).NotTo(HaveOccurred())
+			ctrl.startWatches = []watchDescription{{
+				src: source.NewKindWithCache(&appsv1.Deployment{}, c),
+			}}
+
+			By("running the cache and waiting for it to sync")
+			go func() {
+				defer GinkgoRecover()
+				Expect(c.Start(context.TODO())).To(Succeed())
+			}()
+			close(done)
+		})
+
+		It("should error when timeout is set to a very low value such that cache cannot sync", func(done Done) {
+			ctrl.CacheSyncTimeout = 1 * time.Nanosecond
+			c, err := cache.New(cfg, cache.Options{})
+			Expect(err).NotTo(HaveOccurred())
+			ctrl.startWatches = []watchDescription{{
+				src: source.NewKindWithCache(&appsv1.Deployment{}, c),
+			}}
+
+			By("running the cache and waiting for it to sync")
+			go func() {
+				defer GinkgoRecover()
+				err = ctrl.Start(context.TODO())
```

Comment:
`Start` seems to return a closed channel, so I haven't cancelled the context later. Can I cancel the context after `Start` fails or succeeds, or would it be better to use a wrapper that closes the channel as you have mentioned previously.

Reply:
This will pass with any timeout, because the cache is never started which short-circuits the actual `WaitForCacheSync`: https://github.com/kubernetes-sigs/controller-runtime/blob/33f1d80ebeaa33c9e3d86c86c713ac5847dbe859/pkg/cache/internal/deleg_map.go#L77-L86

Please write two identical tests that only differ in the timeout.

> Start seems to close the channel, so I haven't cancelled the context later. Can I cancel the context after Start fails or succeeds, or would it be better to use a wrapper that closes the channel as you have mentioned previously.

What channel are you referring to that gets closed by `Start`?

Right now in most cases the `go func` will execute after the test is ended via the `close(done`) in line 176, you would need to block before that via a channel that gets closed at the end of the `go func`. But its unclear to my why this is even in a distinct goroutine.